No Mongo : GitHUb Action to Check Unused Dependencies

shivasankaran18 opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
1. Removed Unused Dependencies and Added Missing Dependencies 
2. Added the Depcheck check to husky(pre-commit) and  the GitHub Action
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:** 3932

Fixes #3932 

**Snapshots/Videos:**
![image](https://github.com/user-attachments/assets/e55c4530-7544-497e-8405-bc3d429036a1)


<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**
NO

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

## Checklist

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass


**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
  - Introduced automated checks for unused and missing dependencies in pull requests.
  - Updated pre-commit hook to verify dependencies before committing changes.
  - Added configuration to customize which files and packages are ignored during dependency checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shivasankaran18 commented on .depcheckrc.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "ignores": [
+    "GraphQl",
+    "screens",
+    "subComponents",
+    "setup",
+    "state",
+    "utils",
+    "style",
+    "components",
+    "assets",
+    "Constant",
+    "types",
+    "expect",
```
@coderabbitai no making this change does not make Depcheck to ignoreDirs .. Depcheck only does this for those listed under ignores

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -544,3 +544,26 @@ jobs:
+
+
+
+
```
resolved @coderabbitai

shivasankaran18 commented:
@coderabbitai full review

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -543,4 +543,23 @@ jobs:
+        run: npm install
+
+      - name: Check for Unused and Missing Dependencies
+        run: npm run check-dependencies
```
resolved @coderabbitai

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -543,4 +543,22 @@ jobs:
+        run: npm install
+
+      - name: Check for Unused and Missing Dependencies
+        run: npm run check-dependencies
```
fixed @coderabbitai

shivasankaran18 commented:
@palisadoes  could you review the PR ?

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -543,4 +543,27 @@ jobs:
         run: |
           source venv/bin/activate
-          python .github/workflows/scripts/check_docstrings.py --directories .github
+          python .github/workflows/scripts/check_docstrings.py --directories .github     
```
which trailing spaces @coderabbitai..what do you mean ?

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -543,4 +543,27 @@ jobs:
+        run: npm ci
+
+      - name: Check for Unused and Missing Dependencies
+        run: npm run check-dependencies
```
resolved @coderabbitai

shivasankaran18 commented:
@palisadoes sir could you review the PR ?

palisadoes reviewed (changes requested):
There are missing dependencies, why does your test pass?

shivasankaran18 commented:
which missing dependecnies sir? I have added checks that looks for missing dependencies too? Could not get u sir @palisadoes

palisadoes commented:
1. I ran the command and there are missing and unused dependencies.
2. The command exits with a 255 code, not zero.

If these conditions are true, why does your check pass?

![image](https://github.com/user-attachments/assets/01d62362-de35-4f97-86b3-783b6e714dfa)

palisadoes reviewed (changes requested):
1. I ran the command and there are missing and unused dependencies.
2. The command exits with a 255 code, not zero.

If these conditions are true, why does your check pass?

![image](https://github.com/user-attachments/assets/01d62362-de35-4f97-86b3-783b6e714dfa)

shivasankaran18 commented:
@coderabbitai full review

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -560,4 +560,27 @@ jobs:
         run: |
           source venv/bin/activate
-          python .github/workflows/scripts/check_docstrings.py --directories .github
+          python .github/workflows/scripts/check_docstrings.py --directories .github   
```
resolved

shivasankaran18 commented on .husky/pre-commit:
```diff
@@ -1,9 +1,9 @@
 npm run lint-staged
 npm run typecheck
 npm run update:toc
+npm run check-dependencies
```
resolved

shivasankaran18 commented on .husky/pre-commit:
```diff
@@ -1,9 +1,9 @@
+npm run check-dependencies
 
-git add .
+git add .
```
resolved

shivasankaran18 commented on .github/workflows/pull-request.yml:
```diff
@@ -561,3 +561,26 @@ jobs:
         run: |
           source venv/bin/activate
           python .github/workflows/scripts/check_docstrings.py --directories .github
+  Check-Dependencies:
+    name: Check Dependencies
+    runs-on: ubuntu-latest
+    needs: [Code-Quality-Checks]
+    steps:
+      - name: Checkout the Repository
+        uses: actions/checkout@v4
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '22.x'
+
+      - name: Cache npm modules
+        uses: actions/cache@v4
+        with:
+          path: ~/.npm
+          key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json', 'package.json') }}
+      - name: Install Dependencies
+        run: npm ci
+
+      - name: Check for Unused and Missing Dependencies
+        run: npm run check-dependencies
```
resolved

shivasankaran18 commented:
![image](https://github.com/user-attachments/assets/b9d38eb9-6c96-4ffe-9288-1eb05a8352d9)

@palisadoes  sir it works well on my local.. I have updated the code to the  upstream and updated the missing dependencies too .. could you review the PR now ?

shivasankaran18 commented:
@palisadoes  sir , Depcheck gives some false positives .These seem to be caused by the way path aliases are handled in the project.. Eg : an import like import Users from 'screens/Users' (defined via tsconfig.json paths) is flagged as a missing dependency (screens), even though it points to a local file

Hence I have added all the sub directories to the ignores 

Then , depcheck only detects imports/usage in source files. It does not recognize:
- CLI tools (e.g. markdown-toc)

- Config-based plugins (e.g. typedoc-plugin-markdown)

- Indirect usages (e.g. CSS mocks like identity-obj-proxy)

eg : `const command = `markdown-toc -i "${file}" --bullets "-"`;
`
`"plugin": ["typedoc-plugin-markdown"]
`

Some packages like babel-jest, @types/jest, and eslint-plugin-jest are temporarily required while the codebase still contains Jest tests.

palisadoes commented:
Would limiting the check to specific folders fix the issue?

shivasankaran18 commented:
@palisadoes Could you review the PR when you find time?

palisadoes commented:
Closing as abandoned
